### PR TITLE
[Hexagon] RzIL uplifiting

### PR DESCRIPTION
**Pull request by Rot127**
# SQUASH ME

commit message: Uplift Hexagon architecture to RzIL

The general structure is, that every (sub-)instruction has a getter for it's RzIL code.
Calling the getter will return the RzIL operation.
If RzIL for an instruction is requested, the plugin makes a decision. Because Hexagon only executes whole instruction packets. If the instruction is not the last instruction in a packet, it will simply return `EMPTY()`. If the RzIL for the last instruction in a packet is requested, it will get the RzIL operations for all instructions in the packet, shuffles them into the correct execution order (according to some rules) and returns the complete operation for the packet.

The RzIL code was entirely generated with the [rzil-compiler](https://github.com/Rot127/rzil-compiler/), using the semantic definition of the [QEMU Hexagon module](https://github.com/qemu/qemu/tree/master/target/hexagon).

Currently successful compile instructions (and tested):
```
[*] 1581/1733 standard instructions compiled.
[*] 431/643 HVX instructions compiled.
[*] In total: 2012/2376 instructions compiled.
```

It was tested with:
- (Semantic tests) `rz-tracetest` against the execution trace of the QEMU Hexagon test binaries.
- (Bug free and semi-semtantic test) Adding tests which simply execute the test binaries to ensure leak and segfault free execution. Also it is executed until a certain instruction is reached (end of `main` or `loc.pass` symbol), partially testing it executes correctly.

For the uplifting several changes and modernization had to be made:

- Enhance consistency of decoding
  - Allow to disassemble an instruction without copying the result. This is used if the given buffer of instruction bytes is larger than one instruction width. In this case, as many instructions as the buffer can hold are disassembled and buffered for later.
  - Generally enhance buffering of instructions.
  - Allow to mark a packet as valid before it is completely decoded (in case we know it must be valid, e.g. if it is a jump target of a valid packet).
- Fix (hopefully) all memory leaks of the Hexagon plugin.
- Changes to register getters, because RzIL needs finer control to translate alias or explicit register names to their real register.
  - Getter for register name is now done by table, so for future distinction between DSP version we can just select another table.
  - Translation functions from register alias or explicit name to their real register.
  - Each operand contains now it's variable ID (e.g. `d` for register `Rd`) as in the ISA (for mapping in the RzIL code).
- Ease debugging by tracking in more precision, if an instruction is added to a stale, active or new packet.
- Add registers `C20` - `C29` (not yet present in LLVM)
- Some renaming to make the code more readable.

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Hexagon uplifting PR.

TODO after merge

- [x] `rz-rzilcompiler`
  - [x] Transfer ownership to `RizinOrg`
  - [x] Tag `rz-rzilcompile` as `v1.0`
- [ ] `rz-hexagon`
  - [x] Change submodule to use `RizinOrg` repo
  - [x] Use tag of `rz-rzilcompiler`
  - [x] Refactor `rz-hexagon` to build as proper python package.
  - [ ] Tag `rz-hexagon` as `v1.1`

**Test plan**

Added, all green

### CODE REVIEW

**Comment by Rot127** [comment]
All tests pass. I won't mark it as read yet, because I still want to implement the float instructions.

Also, is there a way to run a bash script as a test?

@XVilka @thestr4ng3r @wargio 
But could you please take a look at this: https://github.com/rizinorg/rizin/pull/3837/commits/56769c752f73cffcc673ada790f7f73cd398a441

**Comment by Rot127** [comment]
Open this for the first round of review now.

There are still the `rz-test` cases missing. Due to inner working of the RzIL generation it is not really possible to add the `asm` style tests (so I check it again to be sure). Instead I would simply execute most [test bins](https://github.com/rizinorg/rizin-testbins/pull/127) to the address where they can be considered passed.

Currently the tests fail due to missing https://github.com/rizinorg/rizin/pull/3973

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 1]
```diff
@@ -0,0 +1,783 @@
...
+#include <rz_types.h>
+#include <rz_util/rz_assert.h>
+#include <rz_il/rz_il_opcodes.h>
+#include <rz_il/rz_il_opbuilder_begin.h>
+
+/**
+ * \brief Sends the IL op at \p start to the position \p newloc.
+ *
+ * Note: THis is a copy of the same function implemented by Qualcomm in QEMU.
+ * See: https://github.com/quic/qemu/blob/d48125de38f48a61d6423ef6a01156d6dff9ee2c/target/hexagon/decode.c#L224-L244
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Was it merged into the mainstream QEMU? Maybe use the official GitLab then?

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 2]
```diff
@@ -0,0 +1,783 @@
...
+		rz_vector_assign_at(ops, i + direction, &tmp_op);
+	}
+}
+
+/**
+ * \brief Shuffles the IL operations of the packet instructions into the correct execution order
+ * and stores the result in \p p->il_ops
+ *
+ * The shuffle algorithm implemented here is a copy of Qualcomm's implementation in QEMU:
+ * https://github.com/quic/qemu/blob/d48125de38f48a61d6423ef6a01156d6dff9ee2c/target/hexagon/decode.c#L423-L543
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Same

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+		// Incomplete packets cannot be executed.
+		return false;
+	}
+	if (rz_vector_empty(p->il_ops)) {
+		RZ_LOG_WARN("Valid packet without RZIL instructions encountered! pkt addr = 0x%" PFMT32x "\n", p->pkt_addr);
+		return false;
+	}
+	RzVector *ops = p->il_ops;
+
+	//
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Can be just one line, I think?

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 4]
```diff
@@ -0,0 +1,783 @@
...
+				/*
+				 * Don't set flag, since we don't want to shuffle past
+				 * a .new value
+				 */
+			} else {
+				flag = true;
+			}
+		}
+
+		if (changed) {
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Probably move these 3 lines one position upper? And put a new line after that

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+			continue;
+		}
+		/* Compares go first, may be reordered with regard to each other */
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Comparisons?

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+		for (flag = false, i = 0; i < last_insn + 1; i++) {
+			op = rz_vector_index_ptr(ops, i);
+			if ((op->attr & HEX_IL_INSN_ATTR_WPRED) &&
+				(op->attr & HEX_IL_INSN_ATTR_MEM_WRITE)) {
+				/* This should be a compare (not a store conditional) */
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: comparison?

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+			// Insn operation 1
+			if (cur_il_insn->op1.attr != HEX_IL_INSN_ATTR_INVALID) {
+				cur_il_insn->op1.hi = pos->bin.insn;
+				rz_vector_push(p->il_ops, &cur_il_insn->op1);
+			}
+		}
+	}
+	return true;
+not_impl:
+	RZ_LOG_INFO("Hexagon instruction %" PFMT32d " not implemented.\n", pos->bin.insn->identifier);
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Maybe WARNING instead?

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 8]
```diff
@@ -0,0 +1,783 @@
...
+ * \param get_pkt_op If true, it returns the IL operation of the whole packet at \p addr.
+ * It will return EMPTY() if there is no packet which starts at \p addr.
+ * If false, the behavior is as documented above.
+ * \return RzILOpEffect* Sequence of operations to emulate the packet.
+ */
+RZ_IPI RzILOpEffect *hex_get_il_op(const ut32 addr, const bool get_pkt_op) {
+	static bool might_has_jumped = false;
+	HexState *state = hexagon_get_state();
+	if (!state) {
+		RZ_LOG_WARN("Could not get hexagon state!\n");
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Maybe "hexagon instruction decoder state" or something like that? I think just "hexagon state" is a bit misleading.

File: librz/analysis/arch/hexagon/hexagon_il.c [hunk 9]
```diff
@@ -0,0 +1,783 @@
...
+}
+
+RZ_IPI RZ_OWN RzILOpEffect *hex_cancel_slot(RZ_BORROW HexPkt *pkt, ut8 slot) {
+	rz_return_val_if_fail(pkt, NULL);
+	if (slot > 3) {
+		rz_warn_if_reached();
+		RZ_LOG_WARN("Slot %d does not exist!", slot);
+	}
+	pkt->il_op_stats.slot_cancelled |= (1 << slot);
+	// TODO Does this really need to be tracked in the VM?
```
Comment (XVilka) [inline librz/analysis/arch/hexagon/hexagon_il.c]: Likely not. Depends on how it affects the actual eexecution

File: librz/analysis/arch/ppc/ppc_il.h [hunk 10]
```diff
@@ -23,15 +23,6 @@
  */
 #define DCACHE_LINE_SIZE 128
 
-/**
```
Comment (XVilka) [inline librz/analysis/arch/ppc/ppc_il.h]: Why this? Is it unused? Please send as a separate PR then please, so we can remove it from this PR

File: librz/asm/arch/hexagon/hexagon.c [hunk 11]
```diff
@@ -20,848 +26,394 @@
 #include "hexagon.h"
 #include "hexagon_insn.h"
 #include "hexagon_arch.h"
+#include "hexagon_reg_tables.h"
+RZ_API ut32 hex_resolve_reg_enum_id(HexRegClass class, ut32 reg_num) {
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.c]: Missing empty line. Please also use a proper name prefix if it's API. Also brief doxygen description.

File: librz/asm/arch/hexagon/hexagon.c [hunk 12]
```diff
@@ -20,848 +26,394 @@
...
-		return get_alias ? "M1" : "C7";
-	case HEX_REG_CTR_REGS_C8:
-		return get_alias ? "USR" : "C8";
+		return reg_num;
+	}
 	}
+	rz_warn_if_reached();
+	return UT32_MAX;
+}
+const char *hex_get_ctr_regs(int reg_num, bool get_alias, bool get_new, bool reg_num_is_enum) {
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.c]: Missing empty line

File: librz/asm/arch/hexagon/hexagon.c [hunk 13]
```diff
@@ -922,3 +474,61 @@ int resolve_n_register(const int reg_num, const ut32 addr, const HexPkt *p) {
...
+/**
+ * \brief Returns a HexOp of the given register number and class.
+ *
+ * \param reg_num The register number as in the name.
+ * \param reg_class The HexRegClass this register belongs to.
+ * \param tmp_reg Flag if the register is a .new register.
+ *
+ * \return A setup HexOp. Currently the HexOp.attr field is *not* set!
+ */
+RZ_API const HexOp hex_explicit_to_op(ut32 reg_num, HexRegClass reg_class, bool tmp_reg) {
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.c]: Please use proper naming

File: librz/asm/arch/hexagon/hexagon.c [hunk 14]
```diff
@@ -922,3 +474,61 @@ int resolve_n_register(const int reg_num, const ut32 addr, const HexPkt *p) {
...
+
+/**
+ * \brief Returns a HexOp of the given register alias.
+ *
+ * \param alias The alias to get the HexOp for.
+ * \param tmp_reg Flag if the alias is referring to the .new register.
+ *
+ * \return A setup HexOp. Currently the HexOp.attr field is *not* set!
+ */
+RZ_API const HexOp hex_alias_to_op(HexRegAlias alias, bool tmp_reg) {
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon.c]: Etc

File: librz/asm/arch/hexagon/hexagon_arch.c [hunk 15]
```diff
@@ -77,7 +84,7 @@ static inline bool is_endloop01_pkt(const ut8 pb_hi_0, const ut8 pb_hi_1) {
  * \param addr The address of the instruction.
  * \return Pointer to instruction or NULL if none was found.
  */
-static HexInsnContainer *hex_get_hic_at_addr(HexState *state, const ut32 addr) {
+RZ_API HexInsnContainer *hex_get_hic_at_addr(HexState *state, const ut32 addr) {
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon_arch.c]: Same

File: librz/asm/arch/hexagon/hexagon_arch.c [hunk 16]
```diff
@@ -103,6 +111,78 @@ static inline bool hic_at_addr(RZ_NONNULL const HexInsnContainer *hic, const ut3
...
+ * \brief Gives for an ISA register character the register name.
+ * E.g.: If the ISA instruction uses the variable "Rd", it passes 'd' as identifier to this function.
+ * The function returns a concrete register name like "R3", "R10" or any other name which is associated with the id.
+ *
+ * \param hi The hexagon instruction.
+ * \param isa_id The ISA register character.
+ * \param new_reg If true it will return the .new register name ("R3_tmp", "R10_tmp" etc.)
+ * \return const char * The concrete register name. Or NULL on error.
+ */
+RZ_API const HexOp *hex_isa_to_reg(const HexInsn *hi, const char isa_id, bool new_reg) {
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon_arch.c]: Same

File: librz/asm/arch/hexagon/hexagon_arch.c [hunk 17]
```diff
@@ -646,24 +752,100 @@ static HexInsnContainer *hex_to_new_pkt(HexState *state, const HexInsnContainer
...
+		if (pkt->last_access < min_time) {
+			min_time = pkt->last_access;
+			oldest = i;
+		}
+		if (pkt->last_access > max_time) {
+			max_time = pkt->last_access;
+			newest = i;
+		}
+	}
+	RZ_LOG_DEBUG("╭─────┬──────────────┬─────┬──────────────────┬───────────────╮\n");
```
Comment (XVilka) [inline librz/asm/arch/hexagon/hexagon_arch.c]: We should have general API for that, inside RzTable.

File: librz/asm/p/asm_hexagon.c [hunk 18]
```diff
@@ -111,6 +117,22 @@ static RZ_OWN RzPVector /*<RzAsmTokenPattern *>*/ *get_token_patterns() {
 	return pvec;
 }
 
+static void compile_token_patterns(RZ_INOUT RzPVector /*<RzAsmTokenPattern *>*/ *patterns) {
+	rz_return_if_fail(patterns);
+
+	void **it;
+	rz_pvector_foreach (patterns, it) {
+		RzAsmTokenPattern *pat = *it;
+		if (!pat->regex) {
```
Comment (XVilka) [inline librz/asm/p/asm_hexagon.c]: You can reverse the condition and use `continue` to reduce nesting

File: librz/il/il_routines.c [hunk 19]
```diff
@@ -0,0 +1,125 @@
+// SPDX-FileCopyrightText: 2023 Rot127 <unisono@quyllur.org>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_il/rz_il_opcodes.h>
+#include <rz_types.h>
+
```
Comment (XVilka) [inline librz/il/il_routines.c]: Add `\file` describing the purpose of these functions (helpers to form bitwise operations)

File: librz/il/il_routines.c [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+	RzILOpPure *op_AND_13 = rz_il_op_new_log_and(rz_il_op_pure_dup(t), rz_il_op_new_cast(32, rz_il_op_new_b0(), rz_il_op_new_bitv_from_st64(32, 0xff0000)));
+	RzILOpPure *op_RSHIFT_15 = rz_il_op_new_shiftr(rz_il_op_new_b0(), op_AND_13, rz_il_op_new_bitv_from_st64(32, 8));
+	RzILOpPure *op_OR_16 = rz_il_op_new_log_or(op_OR_10, op_RSHIFT_15);
+	RzILOpPure *op_AND_19 = rz_il_op_new_log_and(rz_il_op_pure_dup(t), rz_il_op_new_cast(32, rz_il_op_new_b0(), rz_il_op_new_bitv_from_st64(32, 0xff000000)));
+	RzILOpPure *op_RSHIFT_21 = rz_il_op_new_shiftr(rz_il_op_new_b0(), op_AND_19, rz_il_op_new_bitv_from_st64(32, 24));
+	RzILOpPure *op_OR_22 = rz_il_op_new_log_or(op_OR_16, op_RSHIFT_21);
+	return op_OR_22;
+}
+
+RZ_OWN RzILOpPure *rz_il_bswap64(RZ_BORROW RzILOpPure *t) {
```
Comment (XVilka) [inline librz/il/il_routines.c]: Please add Doxygen for all these functions, lets keep RzIL implementation well-documented.

File: librz/include/rz_il/rz_il_opbuilder_begin.h [hunk 21]
```diff
@@ -166,4 +166,19 @@
 #define JMP(tgt)        rz_il_op_new_jmp(tgt)
 #define GOTO(lbl)       rz_il_op_new_goto(lbl)
 
+#define NOT_IMPLEMENTED \
```
Comment (XVilka) [inline librz/include/rz_il/rz_il_opbuilder_begin.h]: I recommend extracting this change in the IL code itself into a separate PR.

**Review by XVilka — CHANGES_REQUESTED** [review]
The code looks great. Amazing job!

File: librz/asm/arch/hexagon/hexagon.c (hunk 11 above)
Comment (Rot127) [reply librz/asm/arch/hexagon/hexagon.c]: Proper naming means `hexagon_...`? If yes, those are a lot of changes. I would like to move this to another PR then.

Comment (Rot127) [reply librz/asm/arch/hexagon/hexagon.c]: Ah, also. Those are techincally `IPI`s. But because the analsysis part relies so heavily on the `asm` part, it needs to be `API` so linking works. So it is not "really" API. After RzArch it will have the visibility of IPI again.

Comment (XVilka) [reply librz/asm/arch/hexagon/hexagon.c]: Ok. Then it's fine as is for now. Empty line is still necessary though...

File: librz/asm/arch/hexagon/hexagon_arch.c (hunk 17 above)
Comment (Rot127) [reply librz/asm/arch/hexagon/hexagon_arch.c]: Couldn't do the last column which is split into 4, which makes it easier to read IMHO. Do you think it is still better? Then I change it. Though it is only debug printing.

**Comment by Rot127** [comment]
Waiting for https://github.com/rizinorg/rizin/pull/3980

File: librz/asm/arch/hexagon/hexagon_arch.c (hunk 17 above)
Comment (XVilka) [reply librz/asm/arch/hexagon/hexagon_arch.c]: It was just a remark, not a requirement. But it would be nice to figure out what's missing in RzTable.

**Comment by Rot127** [comment]
In order that the tests succeed we need https://github.com/rizinorg/rizin-testbins/pull/127 merged though.
Before that maybe look at https://github.com/rizinorg/rizin/pull/3837/files#diff-a1583f59ca1cc7b3c25cde35c24fade6df510112bcf21174d5d2995a5baa14e2

**Comment by Rot127** [comment]
@thestr4ng3r To your notice: https://github.com/rizinorg/rizin/pull/3837/commits/722e5c0a624136b7a07cd4d2ee42e54d3c5fc0c6

**Comment by wargio** [comment]
rebase.

File: test/db/rzil/hexagon [hunk 22]
```diff
@@ -0,0 +1,389 @@
...
+
+# NAME=Run load_align
+# FILE=bins/elf/hexagon/rzil/load_align
+# CMDS=<<EOF
+# aezi
+# aezsu loc.pass
+# EOF
+# EXPECT=<<EOF
+# EOF
+# RUN
```
Comment (wargio) [inline test/db/rzil/hexagon]: ?

File: test/db/rzil/hexagon [hunk 23]
```diff
@@ -0,0 +1,389 @@
...
+NAME=Run test_dotnew
+FILE=bins/elf/hexagon/rzil/test_dotnew
+TIMEOUT=5
+CMDS=<<EOF
+e io.cache=1
+aezi
+aezsu loc.pass
+EOF
+EXPECT=<<EOF
+EOF
```
Comment (wargio) [inline test/db/rzil/hexagon:215]: should we expect some output somewhere? or maybe a register dump after the execution?

File: test/db/rzil/hexagon (hunk 22 above)
Comment (Rot127) [reply test/db/rzil/hexagon]: It executes until it hits the `pass` location. If it fails the test will time out since it never reaches it. Basically the poor mans attempt to check if everything was fine. Since we cannot reasonably run rz-tracetest.

File: test/db/rzil/hexagon (hunk 23 above)
Comment (Rot127) [reply test/db/rzil/hexagon:215]: Indeed. A register dump would be good. Let me add this later.

File: test/db/rzil/hexagon (hunk 22 above)
Comment (wargio) [reply test/db/rzil/hexagon]: yes but this is commented. (thus not executed)

Comment (Rot127) [reply test/db/rzil/hexagon]: Ouh, yes. Just tested it. Something is wrong there and I forgot about it.

File: test/db/rzil/hexagon (hunk 23 above)
Comment (Rot127) [reply test/db/rzil/hexagon:215]: On a second thought, it might not be too useful. The register dump would never be printed in case of a timeout. And if everything executed correctly, it says pretty much nothing interesting anyways.

Would you agree?

**Comment by Rot127** [comment]
If this is green, could you guys take a look at it again.

**Comment by XVilka** [comment]
> If this is green, could you guys take a look at it again.

It's not:
```
101/102 rizin:unit / tokens               FAIL            1.03s   killed by signal 6 SIGABRT
>>> UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MALLOC_PERTURB_=183 LD_LIBRARY_PATH=/home/runner/work/rizin/rizin/build/librz/egg:/home/runner/work/rizin/rizin/build/librz/reg:/home/runner/work/rizin/rizin/build/librz/core:/home/runner/work/rizin/rizin/build/librz/hash:/home/runner/work/rizin/rizin/build/librz/bp:/home/runner/work/rizin/rizin/build/librz/search:/home/runner/work/rizin/rizin/build/librz/asm:/home/runner/work/rizin/rizin/build/librz/demangler:/home/runner/work/rizin/rizin/build/librz/flag:/home/runner/work/rizin/rizin/build/librz/parse:/home/runner/work/rizin/rizin/build/librz/debug:/home/runner/work/rizin/rizin/build/librz/crypto:/home/runner/work/rizin/rizin/build/librz/io:/home/runner/work/rizin/rizin/build/librz/il:/home/runner/work/rizin/rizin/build/librz/util:/home/runner/work/rizin/rizin/build/librz/magic:/home/runner/work/rizin/rizin/build/librz/lang:/home/runner/work/rizin/rizin/build/librz/cons:/home/runner/work/rizin/rizin/build/librz/bin:/home/runner/work/rizin/rizin/build/librz/sign:/home/runner/work/rizin/rizin/build/librz/type:/home/runner/work/rizin/rizin/build/librz/diff:/home/runner/work/rizin/rizin/build/librz/config:/home/runner/work/rizin/rizin/build/librz/syscall:/home/runner/work/rizin/rizin/build/librz/socket:/home/runner/work/rizin/rizin/build/librz/analysis:/home/runner/work/rizin/rizin/build/librz/main:/home/runner/lib/x86_64-linux-gnu:/home/runner/lib:/home/runner/lib64: /home/runner/work/rizin/rizin/build/test/unit/test_tokens
```

**Comment by Rot127** [comment]
Going to move https://github.com/rizinorg/rizin/pull/3837/commits/458fc61836e82da8ceee5de36c58dca6d38bd200 into a separated PR. But first wanted to see if it finally works.

**Comment by Rot127** [comment]
So there is still an error for the `v68_scalar` binary. Some dwarf issue. Ignored it for now, since it is out of scope of this PR though But besides this, please take another look.

**Comment by Rot127** [comment]
So besides https://github.com/rizinorg/rizin/issues/4121 everything runs fine now. I have no idea what the problem with it is unfortunately. But it isn't related to the RzIL stuff.

The appveyor fails due to https://github.com/rizinorg/rizin/issues/4121 and some timeouts, as it looks like. Though I wouldn't like to increase them anymore since they are already pretty high.

File: librz/arch/asm.c [hunk 24]
```diff
@@ -1861,6 +1861,18 @@ rz_asm_colorize_asm_str(RZ_BORROW RzStrBuf *asm_str, RZ_BORROW RzPrint *p, RZ_NU
...
+ * \brief Free a RzAsmParseParam
+ *
+ * \param p The parameter struct.
+ */
+RZ_API void rz_asm_parse_param_free(RZ_OWN RzAsmParseParam *p) {
+	if (!p) {
+		return;
+	}
+	free(p);
+}
```
Comment (wargio) [inline librz/arch/asm.c]: maybe just have in the headers `#define rz_asm_parse_param_free(x) free(x)`

Comment (Rot127) [reply librz/arch/asm.c]: Decided against it because it breaks reverse search for clangd (at least in my case).
But thanks for the hint. Found two leaks this way.

File: librz/arch/isa/hexagon/hexagon_arch.c [hunk 25]
```diff
@@ -646,24 +755,101 @@ static HexInsnContainer *hex_to_new_pkt(HexState *state, const HexInsnContainer
...
+			}
+		}
+		rz_strbuf_append(pkt_line, "\n");
+		RZ_LOG_DEBUG("%s", rz_strbuf_get(pkt_line));
+		rz_strbuf_fini(pkt_line);
+		if (i < HEXAGON_STATE_PKTS - 1) {
+			RZ_LOG_DEBUG("├─────┼──────────────┼─────┼──────────────────┼───┼───┼───┼───┤\n");
+		} else {
+			RZ_LOG_DEBUG("╰─────┴──────────────┴─────┴──────────────────┴───┴───┴───┴───╯\n");
+		}
```
Comment (wargio) [inline librz/arch/isa/hexagon/hexagon_arch.c:847]: when `RZ_LOG_DEBUG` id not enabled this is kinda useless. maybe change this as
```c
RZ_LOG_DEBUG(i < HEXAGON_STATE_PKTS - 1 ?
"├─────┼──────────────┼─────┼──────────────────┼───┼───┼───┼───┤\n" :
"╰─────┴──────────────┴─────┴──────────────────┴───┴───┴───┴───╯\n"
);
```

**Review by wargio — APPROVED** [review]
mostly is ok. just a stupid detail for me.

File: librz/arch/asm.c [hunk 26]
```diff
@@ -1861,6 +1861,15 @@ rz_asm_colorize_asm_str(RZ_BORROW RzStrBuf *asm_str, RZ_BORROW RzPrint *p, RZ_NU
...
 }
 
+/**
+ * \brief Free a RzAsmParseParam
+ *
+ * \param p The parameter struct.
+ */
+RZ_API void rz_asm_parse_param_free(RZ_OWN RZ_NULLABLE RzAsmParseParam *p) {
+	free(p);
+}
```
Comment (wargio) [inline librz/arch/asm.c]: i'm not a big fan of this. but if the others are ok, then i'm ok. @XVilka @thestr4ng3r

Comment (Rot127) [reply librz/arch/asm.c]: Yeah, the only reason I kept it, was that we might extend the parse parameters. So the function is mostly there for future extensions. But we can also remove it or do the macro thing if the others agree with you.

File: librz/arch/isa/hexagon/hexagon_arch.c (hunk 25 above)
Comment (Rot127) [reply librz/arch/isa/hexagon/hexagon_arch.c:847]: Yeah, right. Let me add an early return to the function.

This logging is mostly there, because it visualizes nicely how the packets are filled over time. A thing I found really difficult to track just by looking at the debugger variables. This helps with finding some weird instruction buffering issues.

We can also remove it, if you feel it is unnecessary overhead.

**Review by XVilka — APPROVED** [review]
LGTM but it would be nice to extract non-Hexagon changes in a separate PR and merge it first.

**Comment by Rot127** [comment]
Please let me run all `rz-tracetest` tests after https://github.com/rizinorg/rizin/pull/4373. Just to be sure.

**Comment by Rot127** [comment]
Ok, all good now:
```
./run-test-sets.sh -i -t essentials,float
test_vspliceb [PASS]
test_vpmpyh [PASS]
test_vminh [PASS]
test_vmaxh [PASS]
test_vlsrw [PASS]
test_vcmpw [PASS]
test_vcmpb [PASS]
test_vavgw [PASS]
test_round [PASS]
test_reorder [PASS]
test_packet [PASS]
test_mpyi [PASS]
test_lsr [PASS]
test_jmp [PASS]
test_hwloops [PASS]
test_hl [PASS]
test_fibonacci [PASS]
test_ext [PASS]
test_dotnew [PASS]
test_cmp [PASS]
test_clobber [PASS]
test_call [PASS]
test_bitsplit [PASS]
test_bitcnt [PASS]
test_abs [PASS]
usr [PASS]
v68_scalar [PASS]
v73_scalar [PASS]
test-vma [PASS]
load_align [PASS]
multi_result [PASS]
overflow [PASS]
first [PASS]
mem_noshuf [PASS]
preg_alias [PASS]
dual_stores [PASS]
mem_noshuf_exception [PASS]
read_write_overlap [PASS]
reg_mut [PASS]
misc [PASS]
fpstuff [PASS]
```

**Comment by Matheus-Garbelini** [comment]
Hi, just a question. Is the curent dev branch of rizin using latest changes from https://github.com/rizinorg/rz-hexagon and https://github.com/rizinorg/rz-rzilcompiler ?

**Comment by Rot127** [comment]
@Matheus-Garbelini Yes. If you plan doing research on Hexagon firmware you can ping me on our [Mattermost](https://im.rizin.re) as well.
